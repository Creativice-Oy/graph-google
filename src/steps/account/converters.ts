import generateEntityKey from '../../utils/generateEntityKey';
import { entities } from '../../constants';
import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';

interface CreateAccountEntityParams {
  account: {
    googleAccountId: string;
    name: string;
  };
  domainNames: string[];
  primaryDomain?: string;
}

export function getAccountKey(googleAccountId: string) {
  return generateEntityKey(entities.ACCOUNT._type, googleAccountId);
}

export function createAccountEntity(data: CreateAccountEntityParams) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: entities.ACCOUNT._class,
        _key: getAccountKey(data.account.googleAccountId),
        _type: entities.ACCOUNT._type,
        displayName: data.account.name,
        name: data.account.name,
        domains: data.domainNames,
        primaryDomain: data.primaryDomain,
        accountId: data.primaryDomain,
        id: data.account.googleAccountId,
        vendor: 'Google',
      },
    },
  });
}
