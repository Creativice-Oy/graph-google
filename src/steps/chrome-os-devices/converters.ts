import generateEntityKey from '../../utils/generateEntityKey';
import { admin_directory_v1 } from 'googleapis';
import { entities } from '../../constants';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';

export function createChromeOSDeviceEntity(
  data: admin_directory_v1.Schema$ChromeOsDevice,
) {
  const deviceName = data.deviceId as string;

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: generateEntityKey(entities.CHROME_OS_DEVICE._type, deviceName),
        _type: entities.CHROME_OS_DEVICE._type,
        _class: entities.CHROME_OS_DEVICE._class,
        id: deviceName,
        category: 'laptop',
        make: data.model,
        name: deviceName,
        status: data.status,
        model: data.model,
        serial: data.serialNumber,
        version: data.osVersion,
        supportEndDate: data.supportEndDate,
        platformVersion: data.platformVersion,
        firmwareVersion: data.firmwareVersion,
        macAddress: data.macAddress,
        lastSync: parseTimePropertyValue(data.lastSync),
        lastEnrollementTime: parseTimePropertyValue(data.lastEnrollmentTime),
        bootMode: data.bootMode,
      },
    },
  });
}
