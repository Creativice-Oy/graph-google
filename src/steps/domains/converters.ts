import generateEntityKey from '../../utils/generateEntityKey';
import { admin_directory_v1 } from 'googleapis';
import { entities } from '../../constants';
import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';

export function createDomainEntity(data: admin_directory_v1.Schema$Domains) {
  const domainName = data.domainName as string;

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        // Normally we opt not to use the entity _type in the key, but _key
        // cannot be shorter than 10 chars and domain names can be short...
        _key: generateEntityKey(entities.DOMAIN._type, domainName),
        _type: entities.DOMAIN._type,
        _class: entities.DOMAIN._class,
        // The "creationTime" field is a string in UNIX timestamp format
        createdOn: data.creationTime && parseInt(data.creationTime as string),
        id: domainName,
        displayName: domainName,
        name: domainName,
        emailDomain: domainName,
        primary: data.isPrimary === true,
        verified: data.verified === true,
      },
    },
  });
}
