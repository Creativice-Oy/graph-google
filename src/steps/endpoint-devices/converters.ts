import { cloudidentity_v1 } from 'googleapis';
import {
  Entity,
  RelationshipClass,
  createMappedRelationship,
  RelationshipDirection,
} from '@jupiterone/integration-sdk-core';

import { entities, mappedRelationships } from '../../constants';

export function createAccountManagesDeviceMappedRelationship(params: {
  accountEntity: Entity;
  device: cloudidentity_v1.Schema$GoogleAppsCloudidentityDevicesV1Device;
}) {
  return createMappedRelationship({
    _class: RelationshipClass.MANAGES,
    _type: mappedRelationships.ACCOUNT_MANAGES_DEVICE._type,
    _mapping: {
      sourceEntityKey: params.accountEntity._key,
      relationshipDirection: RelationshipDirection.FORWARD,
      skipTargetCreation: false,
      targetFilterKeys: [['_type', 'deviceId']],
      targetEntity: {
        _type: entities.DEVICE._type,
        // for some reason, the deviceId property doesn't exist in the typings, but it's in responses
        deviceId: (params.device as any).deviceId,
      },
    },
  });
}
