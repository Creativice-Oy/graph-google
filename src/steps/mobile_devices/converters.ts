import generateEntityKey from '../../utils/generateEntityKey';
import { admin_directory_v1 } from 'googleapis';
import { entities } from '../../constants';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
  Entity,
  RelationshipClass,
  createDirectRelationship,
} from '@jupiterone/integration-sdk-core';

export function createMobileDeviceEntity(
  data: admin_directory_v1.Schema$MobileDevice,
) {
  const deviceName = data.deviceId as string;
  const email = Array.isArray(data.email) ? data.email[0] : undefined;
  const name = Array.isArray(data.name) ? data.name[0] : undefined;

  return createIntegrationEntity({
    entityData: {
      // The raw data on mobile devices can be very large, so we've removed it.
      source: {},
      assign: {
        _key: generateEntityKey(entities.MOBILE_DEVICE._type, deviceName),
        _type: entities.MOBILE_DEVICE._type,
        _class: entities.MOBILE_DEVICE._class,
        id: deviceName,
        category: 'mobile',
        make: data.model,
        model: data.model,
        serial: data.serialNumber,
        adbStatus: data.adbStatus,
        brand: data.brand,
        deviceCompromisedStatus: data.deviceCompromisedStatus,
        deviceId: data.deviceId,
        email: email,
        encryptionStatus: data.encryptionStatus,
        firstSyncOn: parseTimePropertyValue(data.firstSync),
        hardware: data.hardware,
        hardwareId: data.hardwareId,
        imei: data.imei,
        lastSyncOn: parseTimePropertyValue(data.lastSync),
        manufacturer: data.manufacturer,
        ownerName: name,
        os: data.os,
        status: data.status,
        type: data.type,
        wifiMacAddress: data.wifiMacAddress,
        macAddress: data.wifiMacAddress,
        userAgent: data.userAgent,
      },
    },
  });
}

export function createAccountManagesMobileDeviceRelationship(params: {
  accountEntity: Entity;
  deviceEntity: Entity;
}) {
  return createDirectRelationship({
    _class: RelationshipClass.MANAGES,
    from: params.accountEntity,
    to: params.deviceEntity,
  });
}
