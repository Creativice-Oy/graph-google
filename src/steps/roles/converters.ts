import generateEntityKey from '../../utils/generateEntityKey';
import { entities } from '../../constants';
import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { admin_directory_v1 } from 'googleapis';
import Schema$Role = admin_directory_v1.Schema$Role;

export type RolePrivilegeStrings = {
  privilegeServiceIds: string[];
  privilegeNames: string[];
};

export function getRoleEntityKey(roleId: string | null | undefined) {
  return generateEntityKey(entities.ROLE._type, roleId);
}

export function getRolePrivilegeStrings(
  role: Schema$Role,
): RolePrivilegeStrings {
  return (role.rolePrivileges || []).reduce(
    (privileges, role) => {
      if (role.serviceId) privileges.privilegeServiceIds.push(role.serviceId);
      if (role.privilegeName)
        privileges.privilegeNames.push(role.privilegeName);

      return privileges;
    },
    { privilegeServiceIds: [], privilegeNames: [] } as RolePrivilegeStrings,
  );
}

export function createRoleEntity(role: Schema$Role) {
  const roleId = role.roleId as string;
  const roleName = role.roleName as string;

  const { privilegeServiceIds, privilegeNames } = getRolePrivilegeStrings(role);

  return createIntegrationEntity({
    entityData: {
      source: role,
      assign: {
        _class: entities.ROLE._class,
        _type: entities.ROLE._type,
        _key: getRoleEntityKey(roleId),
        id: roleId,
        name: roleName,
        displayName: roleName,
        description: role.roleDescription,
        systemRole: role.isSystemRole,
        superAdmin: role.isSuperAdminRole,
        privilegeServiceIds,
        privilegeNames,
        kind: role.kind,
        vendor: 'Google',
      },
    },
  });
}
