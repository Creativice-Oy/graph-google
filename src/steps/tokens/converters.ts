import {
  Entity,
  createIntegrationEntity,
  Relationship,
  createDirectRelationship,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { admin_directory_v1 } from 'googleapis';
import { entities } from '../../constants';

export function getTokenKey(data: admin_directory_v1.Schema$Token) {
  return `user:${data.userKey}:client:${data.clientId}`;
}

export function createTokenEntity(data: admin_directory_v1.Schema$Token) {
  const displayName = data.displayText as string;

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: getTokenKey(data),
        _type: entities.TOKEN._type,
        _class: [entities.TOKEN._class],
        clientId: data.clientId,
        name: displayName,
        displayName,
        scopes: data.scopes,
        anonymous: data.anonymous,
        nativeApp: data.nativeApp,
        userKey: data.userKey,
      },
    },
  });
}

export function createUserAssignedTokenRelationship(params: {
  userEntity: Entity;
  tokenEntity: Entity;
}): Relationship {
  return createDirectRelationship({
    _class: RelationshipClass.ASSIGNED,
    from: params.userEntity,
    to: params.tokenEntity,
  });
}
