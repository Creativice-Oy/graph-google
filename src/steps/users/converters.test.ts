import { getMockUser } from '../../../test/mocks';
import { createUserEntity, getCollectionAsFlattendFields } from './converters';

describe('#createUserEntity', () => {
  test('should convert to entity', () => {
    expect(createUserEntity(getMockUser())).toMatchSnapshot();
  });
});

describe('#getCollectionAsFlattenedFields', () => {
  getCollectionAsFlattendFields({
    collection: [
      {
        type: 'work',
      },
    ],
    suffix: 'Address',
    valueMethod: 'formatted',
  });
});
