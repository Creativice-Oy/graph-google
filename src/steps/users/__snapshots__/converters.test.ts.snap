// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createUserEntity should convert to entity 1`] = `
Object {
  "_class": Array [
    "User",
  ],
  "_key": "google_user_123456789",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "agreedToTerms": true,
        "archived": false,
        "changePasswordAtNextLogin": false,
        "creationTime": "2020-08-13T19:03:11.000Z",
        "customerId": "abc123",
        "emails": Array [
          Object {
            "address": "john.doe@jupiterone.io",
            "primary": true,
            "type": "work",
          },
        ],
        "etag": "abcdef",
        "id": "123456789",
        "includeInGlobalAddressList": true,
        "ipWhitelisted": false,
        "isAdmin": false,
        "isDelegatedAdmin": false,
        "isEnforcedIn2Sv": true,
        "isEnrolledIn2Sv": true,
        "isMailboxSetup": true,
        "kind": "admin#directory#user",
        "lastLoginTime": "2020-08-27T00:57:59.000Z",
        "locations": Array [
          Object {
            "area": "desk",
            "buildingId": "RTP",
            "type": "desk",
          },
        ],
        "name": Object {
          "familyName": "Doe",
          "fullName": "John Doe",
          "givenName": "John",
        },
        "orgUnitPath": "/Marketing",
        "organizations": Array [
          Object {
            "customType": "",
            "department": "Marketing",
            "primary": true,
            "title": "CMO",
          },
        ],
        "primaryEmail": "john.doe@jupiterone.io",
        "recoveryEmail": "john.doe.recovery@jupiterone.io",
        "recoveryPhone": "+19999999999",
        "relations": Array [
          Object {
            "type": "manager",
            "value": "john.doe.manager@jupiterone.io",
          },
        ],
        "suspended": false,
      },
    },
  ],
  "_type": "google_user",
  "active": true,
  "agreedToTerms": true,
  "aliases": undefined,
  "archived": false,
  "changePasswordAtNextLogin": false,
  "costCenter": undefined,
  "createdOn": 1597345391000,
  "creationTime": 1597345391000,
  "customType": "",
  "customerId": "abc123",
  "deletionTime": undefined,
  "department": "Marketing",
  "displayName": "John Doe",
  "email": "john.doe@jupiterone.io",
  "firstName": "John",
  "hashFunction": undefined,
  "id": "123456789",
  "includeInGlobalAddressList": true,
  "ipWhitelisted": false,
  "isAdmin": false,
  "isDelegatedAdmin": false,
  "isEnforcedIn2Sv": true,
  "isEnrolledIn2Sv": true,
  "isMailboxSetup": true,
  "kind": "admin#directory#user",
  "lastLoginTime": 1598489879000,
  "lastName": "Doe",
  "managerEmail": "john.doe.manager@jupiterone.io",
  "managerRelation": "john.doe.manager@jupiterone.io",
  "mfaEnabled": true,
  "name": "John Doe",
  "orgUnitPath": "/Marketing",
  "primaryEmail": "john.doe@jupiterone.io",
  "recoveryEmail": "john.doe.recovery@jupiterone.io",
  "recoveryPhone": "+19999999999",
  "role": undefined,
  "suspended": false,
  "suspensionReason": undefined,
  "thumbnailPhotoEtag": undefined,
  "thumbnailPhotoUrl": undefined,
  "title": "CMO",
  "username": "john.doe",
  "workEmail": "john.doe@jupiterone.io",
}
`;
